### DISCUSSION TITLE
`Owl.LiveScreen` terminal input (proof of concept)

### DISCUSSION THREAD

**Discussion by zachallaun** [fuelen/owl#24] [Ideas]
Since it's unclear how far we are from a [portable getch](https://github.com/fuelen/owl/discussions/5), I thought I'd experiment with some way of getting user input while using `Owl.LiveScreen`. 

After some hacking around, I have a proof of concept that I wanted to share:

* [Code](https://github.com/fuelen/owl/compare/main...zachallaun:owl:live-input-proof-of-concept)

![live_input_proof_of_concept](https://user-images.githubusercontent.com/503938/233386451-c7c5e831-9d52-4e34-8348-adb6883c151a.gif)

The basic idea is to allow `Owl.LiveScreen` to render a singleton prompt, and then use `\e7` (cursor position save) and `\e8` (cursor position restore) to maintain the cursor position during renders. The changes I made in the linked branch are _very_ rough -- basically just getting something working. There are a number of open questions:

* How reliable is `\e7` and `\e8`? It seems like most modern terminals support them, but they're non-standard, I believe. There's also `\e[s` and `\e[u` -- there are some notes about them in [this gist](https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#cursor-controls).
* What is the desired API? Right now I've just added an `:on_input` option to `add_block` that can be a function taking the input and block state and returning a new state.
* A limitation with the current API that I ran up against: `LiveScreen.update` replaces the whole state and the owl demo used `Stream.iterate` to move the owl, but I wanted the color to react to the user input. For this proof of concept, I just broke the API and decided that `LiveScreen.update` would merge the maps so that the owl color wouldn't be overriden by every update from the stream iteration. It may be desirable for `LiveScreen.update` to take an update function instead that can be applied to the current state of the block.

**Comment by fuelen** [comment]
Wow, I'm impressed!
* If most of the terminals support these sequences, I think that's enough
* Having `Owl.IO.write(data, state.device)` instead of `Owl.IO.puts(data, state.device)` is a breaking change. Probably, you've already noticed how other examples work with it :) We need something smart in this place
* Any reasons why Process.sleep(70) is replaced with Process.sleep(200)? Are there performance issues?
* A question about API is tricky. As input is global we should probably have a global handler and not per-block.
I think it would be great to have some kind of TEA (The Elm Architecture), but for the time being Owl supports only views from it.  `LiveScreen.update` just sends new data for rendering and I think state management should be on another level.
Maybe, we should try to support input using The Erlang I/O Protocol for `Owl.LiveScreen` and have something like this:
  ```
  input = Owl.IO.input(device: Owl.LiveScreen)
  ```
  and clear the block with input after pressing enter

**Comment by zachallaun** [comment]
> * Having `Owl.IO.write(data, state.device)` instead of `Owl.IO.puts(data, state.device)` is a breaking change. Probably, you've already noticed how other examples work with it :) We need something smart in this place

Yup! For this proof of concept, I really didn’t put any effort towards maintaining compatibility. I think it should be doable though, but the checks needed would have made things more complicated here so I didn’t add it for now. 

> * Any reasons why Process.sleep(70) is replaced with Process.sleep(200)? Are there performance issues?

No reason! It got changed at some point while I was debugging so that I could see what was going on a bit easier and I think I just didn’t change it back. I didn’t exactly stress test it, but I don’t think there are any major performance issues. 

> * A question about API is tricky. As input is global we should probably have a global handler and not per-block.
>   I think it would be great to have some kind of TEA (The Elm Architecture), but for the time being Owl supports only views from it.  `LiveScreen.update` just sends new data for rendering and I think state management should be on another level.

I wonder if a behaviour-driven approach would be most natural here?

```elixir
defmodule MyLiveScreen do
  use Owl.LiveScreen

  @impl true
  def init(_arg, screen) do
    {:ok, screen, input: true}
  end

  @impl true
  def render(screen) do
    %{assigns: assigns} = screen
    # return owl data
  end

  @impl true
  def handle_input(input, screen) do
    # assign stuff, etc. 
    {:noreply, screen}
  end
end
```

This is just a sketch but for more complex terminal UIs, something like this would feel more natural, to me at least. It’s also possible to build something like this on top of the function-based API, so perhaps both could coexist?

>   Maybe, we should try to support input using The Erlang I/O Protocol for `Owl.LiveScreen` and have something like this:
>   ```
>   input = Owl.IO.input(device: Owl.LiveScreen)
>   ```
>     
>   and clear the block with input after pressing enter

Thats another good option! The concern here is the prompt and where it should be located. I’d want to “block out” that space ahead of time so that the entire layout doesn’t shift when I prompt, so you might want some way to declare up front that you’re going to be prompting. 

In this demo as well, I used a static `>` prompt, but it would be nice as well for the user to specify the prompt (as owl data, I’d think).

**Comment by zachallaun** [comment]
Re: API, I actually don't think a behaviour-driven approach is needed, as seen in your own widget implementations. Creating a GenServer yourself to manage state is fine.

**Comment by fuelen** [comment]
> I’d want to “block out” that space ahead of time so that the entire layout doesn’t shift when I prompt, so you might want some way to declare up front that you’re going to be prompting.

The current system with blocks works in a way, that when a block grows in size and shrinks back after then the size of the block remains the same with empty lines, so it doesn't jump.
```elixir
Mix.install([:owl])
Owl.LiveScreen.add_block(:block1, state: "")
Owl.LiveScreen.add_block(:block2, state: "static")

[1, 2, 3, 4, 3, 2, 1]
|> Enum.each(fn lines_number ->
  Owl.LiveScreen.update(:block1, 1..lines_number |> Enum.map_join("\n", &"line#{&1}"))
  Process.sleep(1000)
end)

Owl.LiveScreen.flush()
```

**Comment by zachallaun** [comment]
By the way, only because I’m not sure if it notifies, but I shot you a message on ElixirForum chat 🙂

**Comment by fuelen** [comment]
It notifies :slightly_smiling_face:

**Comment by AlexGx** [comment]
Looks great. Any plans for releasing this feature?
